torch/distributed/_composable/fsdp/fully_shard.py

```diff
@@ -25,15 +26,64 @@ def fully_shard(
+    Constructing multiple groups across the model (e.g. "layer by layer")
+    allows for peak memory savings and communication/computation overlap.
+
+    Implementation-wise, the sharded parameters are represented as
+    :class:`DTensor` s, sharded on dim-0, and the unsharded parameters are
+    represented as :class:`Tensor` s. A module forward pre-hook all-gathers the
+    parameters, and a module forward hook frees them. Similar backward hooks
+    gather parameters and later free parameters/reduce gradients.
+
     Args:
-        mesh (Optional[DeviceMesh]): This mesh defines the sharding and device.
-            If this is a 1D mesh, then this fully shards across the 1D mesh
-            (i.e. FSDP). If this is a 2D mesh, then this shards across the 0th
-            dimension and replicates across the 1st dimension (i.e. HSDP).
-            FSDP/HSDP uses the device given by the mesh's device type. For CUDA
-            or CUDA-like devices, FSDP uses the current device.
+        mesh (Optional[DeviceMesh]): This data parallel mesh defines the
+            sharding and device. If 1D, then parameters are fully sharded
+            across the 1D mesh (FSDP). If 2D, then parameters are sharded
+            across the 0th dim and replicated across the 1st dim (HSDP). The
+            mesh's device type gives the device type used for communication;
+            if a CUDA or CUDA-like device type, then we use the current device.
+        reshard_after_forward (Union[bool, int]): This controls the parameter
+            behavior after forward and can trade off memory and communication:
+            - If ``True``, then this reshards parameters after forward and
+            all-gathers in backward.
+            - If ``False``, then this keeps the unsharded parameters in memory
+            after forward and avoids the all-gather in backward.
+            - If an ``int``, then this represents the world size to reshard to
+            after forward. It should be a number between 1 and the ``mesh``
+            shard dimension size, exclusive. A common choice may be the
+            intra-node size (e.g. ``torch.cuda.device_count()``). This allows
+            the all-gather in backward to be over a smaller world size at the
+            cost of higher memory usage than setting to ``True``.
+            - The root FSDP state has its value specially set to ``False`` as a
+            heuristic since its parameters would typically be immediately
+            all-gathered for backward.
+            - After forward, the parameters registered to the module depend on
+            to this: The registered parameters are the sharded parameters if
+            ``True``; unsharded parameters if ``False``; and the paramters
+            resharded to the smaller mesh otherwise. To modify the parameters
+            between forward and backward, the registered parameters must be the
+            sharded parameters. For ``False`` or an ``int``, this can be done
```

Comment:
Any concrete case where user need to modify the params between fwd/bwd? I would vote for not exposing `reshard` if there's no existing use case

Reply:
I think there are real use cases, but they are uncommon.

Some examples where I think this matters:
- https://github.com/pytorch/pytorch/issues/117421: loading a state dict between forward and backward is silently incorrect
- https://github.com/pytorch/pytorch/issues/117742: modifying parameters without `summon_full_params` for EMA does not persist

The crux of the issue is that the sharded parameters are the source of truth, so any modification must be done to them. Normally, the only modification is the optimizer step; however, there are other valid ones like the state dict load or EMA to be careful of.

If parameters other than the sharded parameters are modified, there is no simple way for FSDP to propagate those changes back to the sharded parameters.